api/segments/serializers.py

```diff
@@ -74,13 +85,34 @@ def validate_project_segment_limit(self, project: Project) -> None:
                 }
             )
 
-    def update(self, instance, validated_data):
-        # use the initial data since we need the ids included to determine which to update & which to create
-        rules_data = self.initial_data.pop("rules", [])
-        self._update_segment_rules(rules_data, segment=instance)
-        # remove rules from validated data to prevent error trying to create segment with nested rules
-        del validated_data["rules"]
-        return super().update(instance, validated_data)
+    def validate_segment_rules_conditions_limit(
+        self, rules_data: dict[str, object]
+    ) -> None:
+        count = self._calculate_condition_count(rules_data)
+        if count > settings.SEGMENT_RULES_CONDITIONS_LIMIT:
+            raise ValidationError(
+                {
+                    "segment": f"The segment has {count} of the maximum condition count "
+                    f"limit of {settings.SEGMENT_RULES_CONDITIONS_LIMIT}."
```

Comment:
I think this could be worded better. Perhaps something like the following:

```suggestion
                    "segment": f"The segment has {count} conditions, which exceeds the maximum "
                    f"condition count of {settings.SEGMENT_RULES_CONDITIONS_LIMIT}."
```

This also triggers a couple of questions: 

 1. Will the FE handle this error correctly? 
 2. Are there segments in the database right now that people will not be able to update due to this limit? (I will check this)

Reply:
Unfortunately, the answer to (2) is yes. We have 57 such segments in the database already. This indicates we may also have issues for self hosted customers too. 

The biggest problem we have is that we can't just tell people 'you'll have to reduce the size of the segment', because that relies on a PUT request... which will fail. 

So, I think we have a few options (whether they are all feasible / worthwhile I'm not sure): 

 1. Only implement the validation on new segments 
 2. Change the logic so that the limit is enforced in the database at a project level so that we can add exceptions for those customers
 3. Add some logic that only validates newly added conditions? 

My preference would be (2) or (3) for sure, but I'm not sure which is most feasible. 

_Query used to obtain data:_ 
```sql
SELECT s.id, count(*)
FROM segments_segment s
LEFT outer JOIN segments_segmentrule sr1 on s.id = sr1.segment_id
left outer join segments_segmentrule sr2 on sr1.id = sr2.rule_id 
left outer join segments_condition sc on sr2.id = sc.rule_id
group by s.id
having count(*) > 100
order by count(*) desc;
```

Reply:
Over a call with @matthewelwell: 

Decided to run a one off migration and include a new model to allow whitelisted segments to evade the limits.